src/plugin.py

```diff
@@ -1,113 +1,264 @@
-            self.minecraft_installation_status = LocalGameState.Installed | LocalGameState.Running
-            self.update_local_game_status(LocalGame('1', LocalGameState.Installed | LocalGameState.Running))
-        if self.minecraft_launcher and not self.local_client.running_process and self.minecraft_installation_status == LocalGameState.Installed | LocalGameState.Running:
-            self.minecraft_installation_status = LocalGameState.Installed
-            self.update_local_game_status(LocalGame('1', LocalGameState.Installed))
-
-    def shutdown(self):
-        pass
+        if self.update_task is None or self.update_task.done():
+            self.update_task = self.create_task(self._update(), "Update Task")
+
+    # Time Tracker
+
+    async def get_game_time(self, game_id, context):
+        try:
+            tracked_time = self.game_time_tracker.get_tracked_time(game_id)
+        except time_tracker.GameNotTrackedException:
+            tracked_time = GameTime(game_id, 0, None)
+        if self._multimc_enabled() and game_id == GameID.Minecraft:
+            multimc_time = self.multimc.get_time()
+        else:
+            multimc_time = GameTime(game_id, 0, None)
+        time = tracked_time.time_played + multimc_time.time_played
+        lastPlayed = utils.compare(tracked_time.last_played_time, multimc_time.last_played_time)
+        log.debug(f"Got game time: {time}")
+        if time == 0 or lastPlayed is None:
+            return None
+        return GameTime(game_id, time, lastPlayed)
+
+    def handshake_complete(self):
+        self.play_time_cache_path = os.path.join(
+            INSTALLED_FOLDER_PATH, "minecraft_play_time_cache.txt"
+        )
+        log.debug(f"Local Play Time Cache Path: {self.play_time_cache_path}")
+        if "game_time_cache" in self.persistent_cache:
+            self.game_time_cache = pickle.loads(
+                bytes.fromhex(self.persistent_cache["game_time_cache"])
+            )
+        else:
+            try:
+                file = open(self.play_time_cache_path, "r")
+                for line in file.readlines():
+                    if line[:1] != "#":
+                        self.game_time_cache = pickle.loads(bytes.fromhex(line))
+                        break
+            except FileNotFoundError:
+                self.game_time_cache = None
+        self.game_time_tracker = time_tracker.TimeTracker(game_time_cache=self.game_time_cache)
+
+    async def shutdown(self):
+        if self.game_time_cache is not None:
+            try:
+                with open(self.play_time_cache_path, "w+") as file:
+                    file.write("# DO NOT EDIT THIS FILE\n")
+                    file.write(self.game_time_tracker.get_time_cache_hex())
+                    log.info("Wrote to local file cache")
+            except time_tracker.GamesStillBeingTrackedException:
+                log.debug("Game time still being tracked. Not setting local cache yet.")

```

Comment:
I did that because it causes an exception otherwise. But I realise now I could just stop tracking the game manually as if the user has finished and then try again. Thanks for noticing that.

Reply:
yes, that would be good workaround to "stop" session even if game is still running when plugin/Galaxy is closed.